packages/devtools_app/test/trees_test.dart

```diff
@@ -119,46 +121,131 @@ void main() {
-      expect(treeNode2.lastChildNodeAtLevel(2), equals(treeNode6));
-      expect(treeNode2.lastChildNodeAtLevel(3), isNull);
+      expect(treeNode3.lastChildNodeAtLevel(0), equals(treeNode3));
+      expect(treeNode3.lastChildNodeAtLevel(1), equals(treeNode5));
+      expect(treeNode3.lastChildNodeAtLevel(2), equals(treeNode9));
+      expect(treeNode3.lastChildNodeAtLevel(3), isNull);
+    });
+
+    test('filterTree', () {
+      final filteredTreeRoots =
+          testTreeNode.filterTree((node) => node.id.isEven);
+      expect(filteredTreeRoots.length, equals(1));
+      final filteredTree = filteredTreeRoots.first;
+      expect(filteredTree.toString(), equals('''
+0
+  2
+  4
+  6
+  8
+'''));
+    });
+
+    test('filterTree when root should be filtered out', () {
+      final filteredTreeRoots =
+          testTreeNode.filterTree((node) => node.id.isOdd);
+      expect(filteredTreeRoots.length, equals(2));
+      final firstRoot = filteredTreeRoots.first;
+      final lastRoot = filteredTreeRoots.last;
+
+      expect(firstRoot.toString(), equals('''
+1
+'''));
+      expect(lastRoot.toString(), equals('''
+3
+  5
+    7
+    9
+'''));
+    });
+
+    test('filterTree when zero nodes match', () {
+      final filteredTreeRoots = testTreeNode.filterTree((node) => node.id > 10);
+      expect(filteredTreeRoots, isEmpty);
+    });
+
+    test('filterTree when all nodes match', () {
+      final filteredTreeRoots = testTreeNode.filterTree((node) => node.id < 10);
+      expect(filteredTreeRoots.length, equals(1));
+      final filteredTree = filteredTreeRoots.first;
+      expect(filteredTree.toString(), equals('''
+0
+  1
+  2
+  3
+    4
+    5
+      6
+      7
+      8
+      9
+'''));
     });
   });
 }
 
-final treeNode0 = TestTreeNode();
-final treeNode1 = TestTreeNode();
-final treeNode2 = TestTreeNode();
-final treeNode3 = TestTreeNode();
-final treeNode4 = TestTreeNode();
-final treeNode5 = TestTreeNode();
-final treeNode6 = TestTreeNode();
-final TreeNode testTreeNode = treeNode0
+final treeNode0 = TestTreeNode(0);
+final treeNode1 = TestTreeNode(1);
+final treeNode2 = TestTreeNode(2);
+final treeNode3 = TestTreeNode(3);
+final treeNode4 = TestTreeNode(4);
+final treeNode5 = TestTreeNode(5);
+final treeNode6 = TestTreeNode(6);
+final treeNode7 = TestTreeNode(7);
+final treeNode8 = TestTreeNode(8);
+final treeNode9 = TestTreeNode(9);
+final TestTreeNode testTreeNode = treeNode0
   ..addAllChildren([
     treeNode1,
-    treeNode2
+    treeNode2,
+    treeNode3
       ..addAllChildren([
-        treeNode3,
-        treeNode4..addAllChildren([treeNode5, treeNode6]),
+        treeNode4,
+        treeNode5
+          ..addAllChildren([
+            treeNode6,
+            treeNode7,
+            treeNode8,
+            treeNode9,
+          ]),
       ]),
   ]);
 
-class TestTreeNode extends TreeNode<TestTreeNode> {}
+class TestTreeNode extends TreeNode<TestTreeNode> {
```

Comment:
you could optionally make this extend DiagnosticableNode to get pretty toString behavior.